### PR TITLE
[SecurityBundle] Allow specifying attributes for `RequestMatcher`

### LINKED ISSUE #45901: [SecurityBundle] Allow specifying attributes._route for RequestMatcher
### Description

Currently, `Symfony\Bundle\SecurityBundle\DependencyInjection\SecurityExtension::createRequestMatcher()` accepts parameter `$attributes`, which I think could be used to specify attributes: {_route: route.id} to match a route instead of path (which I think is more useful for multilingual projects). However, currently it is not possible to actually use this because `Symfony\Bundle\SecurityBundle\DependencyInjection\SecurityExtension::createAuthorization()` (which calls `$this->createRequestMatcher()`) ends with `$access['ips']`, does not use the $attributes parameter at all.

### Example

```yml
security:
   access_control:
      - { attributes: {_route: route_id}, roles: ROLE_USER }
```
or even
```yml
security:
   access_control:
      - { route: route_id, roles: ROLE_USER }
```
(which would require additional code and docs changes)

### PR DESCRIPTION
**Pull request by freiondrej-lmc**
| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #45901
| License       | MIT

The \Symfony\Component\HttpFoundation\RequestMatcher supports array $attributes, which makes it possible to specify a _route (useful e.g. in a multilingual project where $path is translated). However, its current configuration does not offer the possibility to specify the attributes. This PR adds this possibility, so this already existing feature can be leveraged.
I also added a shortcut to just specify "route": "xxx", which translates to "attributes": ["_route": "xxx"].

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Thanks for the PR.

Please rebase and target 6.1 since that's a new feature. Please also update the description of the PR a bit so that ppl that wonder about what the attached patch does can have some clues (eg to start writing the doc about this).
Then, please add a line in the CHANGELOG file of the bundle, and please add a test case if possible :pray:

**Comment by freiondrej-lmc** [comment]
@carsonbot please find me a reviewer

**Comment by chalasr** [comment]
@freiondrej-lmc Thanks for the PR!

> @carsonbot please find me a reviewer

The 6.1 branch is feature-frozen, as such the current focus should be about stabilizing 6.1. So this is for 6.2 and we have 6 months to review and fine-tune it.

**Comment by freiondrej-lmc** [comment]
@chalasr thanks for the information :) just so I know, when was the feature freeze? (so that I can plan better the next time)

**Comment by stof** [comment]
@freiondrej-lmc 2 months before the release, so end of March

**Comment by stof** [comment]
See https://symfony.com/doc/current/contributing/community/releases.html#development

**Review by fabpot — APPROVED** [review]

File: src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/SecurityExtensionTest.php [hunk 1]
```diff
@@ -323,6 +323,105 @@ public function provideAdditionalRequestMatcherConstraints()
...
+        $this->assertSame('add', $call[0]);
+        $args = $call[1];
+        $requestMatcherId = (string) $args[0];
+
+        $requestMatcherDefinition = $container->getDefinition($requestMatcherId);
+        $requestMatcherConstructorArguments = $requestMatcherDefinition->getArguments();
+        $this->assertArrayHasKey(4, $requestMatcherConstructorArguments);
+        $attributes = $requestMatcherConstructorArguments[4];
+        $this->assertArrayHasKey('_route', $attributes);
+        $this->assertEquals('foo_route', $attributes['_route']);
```
Comment (OskarStark) [inline src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/SecurityExtensionTest.php]: assertSame please, if possible

File: src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/SecurityExtensionTest.php [hunk 2]
```diff
@@ -323,6 +323,105 @@ public function provideAdditionalRequestMatcherConstraints()
...
```
Comment (OskarStark) [inline src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/SecurityExtensionTest.php]: Same here

Comment (freiondrej-lmc) [reply src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/SecurityExtensionTest.php]: Thanks for reviewing @OskarStark! 🙂

**Comment by freiondrej-lmc** [comment]
I don't really understand why the tests fail now, when prior to my fixup commit https://github.com/symfony/symfony/pull/45907/commits/4983fd5ccaaa16fa3bf701fb1e89eb23b84fa7ef they passed and I don't see how it could have broken it ... any suggestions would be very welcome 🙂

**Comment by OskarStark** [comment]
> I don't really understand why the tests fail now, when prior to my fixup commit https://github.com/symfony/symfony/commit/4983fd5ccaaa16fa3bf701fb1e89eb23b84fa7ef they passed and I don't see how it could have broken it ... any suggestions would be very welcome 🙂

I think you can ignore the failing test case, its from another PR

**Comment by freiondrej-lmc** [comment]
> > I don't really understand why the tests fail now, when prior to my fixup commit [4983fd5](https://github.com/symfony/symfony/commit/4983fd5ccaaa16fa3bf701fb1e89eb23b84fa7ef) they passed and I don't see how it could have broken it ... any suggestions would be very welcome 🙂
> 
> I think you can ignore the failing test case, its from another PR

@OskarStark 

Well most of it is, but also mine: 
```
1) Symfony\Bundle\SecurityBundle\Tests\DependencyInjection\SecurityExtensionTest::testRegisterAccessControlWithSpecifiedAttributes
[961](https://github.com/symfony/symfony/actions/runs/3098415653/jobs/5016303268#step:8:1006)
Failed asserting that an array has the key 4.
[962](https://github.com/symfony/symfony/actions/runs/3098415653/jobs/5016303268#step:8:1007)

[963](https://github.com/symfony/symfony/actions/runs/3098415653/jobs/5016303268#step:8:1008)
/home/runner/work/symfony/symfony/src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/SecurityExtensionTest.php:360
[964](https://github.com/symfony/symfony/actions/runs/3098415653/jobs/5016303268#step:8:1009)

[965](https://github.com/symfony/symfony/actions/runs/3098415653/jobs/5016303268#step:8:1010)
2) Symfony\Bundle\SecurityBundle\Tests\DependencyInjection\SecurityExtensionTest::testRegisterAccessControlWithSpecifiedRoute
[966](https://github.com/symfony/symfony/actions/runs/3098415653/jobs/5016303268#step:8:1011)
Failed asserting that an array has the key 4.
```
https://github.com/symfony/symfony/actions/runs/3098415653/jobs/5016302933

**Comment by OskarStark** [comment]
Indeed, sorry I just saw the following:
CleanShot 2022-09-21 at 23 11 16@2x

**Comment by freiondrej-lmc** [comment]
@OskarStark I now ran the tests locally and I did not get this error with my code. Maybe there are some changes from a different PR involved? I don’t know the details of the pipeline and how isolated the PRs are from each other 😕 any help will be much appreciated!

**Comment by freiondrej-lmc** [comment]
@OskarStark any ideas would be very welcome. Will this PR even make it into 6.2?

**Review by nicolas-grekas — APPROVED** [review]
(test failures are unrelated)

**Review by welcoMattic — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @freiondrej-lmc.

**Comment by xabbuh** [comment]
@freiondrej-lmc The remaining failures that you mentioned in https://github.com/symfony/symfony/pull/45907#issuecomment-1254038696 were related to the refactoring of the request matchers that happened in the meantime in #47595. #47930 is going to fix them.

**Pull request merged**